### PR TITLE
Virtio console config

### PR DESCRIPTION
**Pull request by fazlamehrab**
Console size feature implementation for the virtio console device.

### CODE REVIEW

File: vmm/src/vm.rs [hunk 1]
```diff
@@ -517,6 +523,9 @@ struct DeviceManager {
 
     // PCI root
     pci: Arc<Mutex<PciConfigIo>>,
+
+    // List the presense of the signals with values from 1 to 64
+    signal_present: [bool; MAX_SIGNAL_VAL + 1],
```
Comment (rbradford) [inline vmm/src/vm.rs]: I think `interested_signals` or something would be a clearer name here

File: vmm/src/vm.rs [hunk 2]
```diff
@@ -1616,6 +1720,41 @@ impl<'a> Vm<'a> {
...
+                signals[index] = sig_num as i32;
+                index += 1;
+            }
+        }
+        if index > 0 {
+            if let Some(console_input) = &self.devices.console_input {
+                let console_input_clone = console_input.clone();
+                set_handler(&signals[0..index], move |received_signals| {
+                    for signal in received_signals.iter() {
+                        if *signal == SIGWINCH {
```
Comment (rbradford) [inline vmm/src/vm.rs]: Could the code for querying and updating be put into console.rs. I see there is similar code in ::new()

File: vmm/src/vm.rs (hunk 1 above)
Comment (rbradford) [reply vmm/src/vm.rs]: Or perhaps `signal_handler_present`.

File: vmm/src/vm.rs [hunk 3]
```diff
@@ -1616,6 +1720,41 @@ impl<'a> Vm<'a> {
         // Unblock all CPU threads.
         vcpu_thread_barrier.wait();
 
+        let mut signals: [i32; MAX_SIGNAL_VAL] = [0; MAX_SIGNAL_VAL];
```
Comment (rbradford) [inline vmm/src/vm.rs]: I think replace this array with a Vec.

File: vmm/src/vm.rs [hunk 4]
```diff
@@ -1616,6 +1720,41 @@ impl<'a> Vm<'a> {
...
+        let mut index = 0;
+        for sig_num in 1..MAX_SIGNAL_VAL {
+            if self.devices.signal_present[sig_num] {
+                signals[index] = sig_num as i32;
```
Comment (rbradford) [inline vmm/src/vm.rs]: And then you can push into the Vec here.

File: vmm/src/vm.rs [hunk 5]
```diff
@@ -1616,6 +1720,41 @@ impl<'a> Vm<'a> {
...
+                index += 1;
```
Comment (rbradford) [inline vmm/src/vm.rs]: Then you don't need `index`

File: vmm/src/vm.rs [hunk 6]
```diff
@@ -1616,6 +1720,41 @@ impl<'a> Vm<'a> {
...
+            }
+        }
+        if index > 0 {
+            if let Some(console_input) = &self.devices.console_input {
+                let console_input_clone = console_input.clone();
+                set_handler(&signals[0..index], move |received_signals| {
```
Comment (rbradford) [inline vmm/src/vm.rs]: And you can just put the reference to the Vec in here which will get turned into a slice.

File: vmm/src/vm.rs [hunk 7]
```diff
@@ -1616,6 +1720,41 @@ impl<'a> Vm<'a> {
...
 
+        let mut signals: [i32; MAX_SIGNAL_VAL] = [0; MAX_SIGNAL_VAL];
+        let mut index = 0;
+        for sig_num in 1..MAX_SIGNAL_VAL {
+            if self.devices.signal_present[sig_num] {
+                signals[index] = sig_num as i32;
+                index += 1;
+            }
+        }
+        if index > 0 {
```
Comment (rbradford) [inline vmm/src/vm.rs]: Just test for Vec emptiness.

File: vmm/src/vm.rs (hunk 3 above)
Comment (rbradford) [reply vmm/src/vm.rs]: In fact if you turned `DeviceManager.signal_present` which is currently an array of booleans into a Vec of ints then you could skip all this work and just pass a reference to that in.

File: vm-virtio/src/console.rs [hunk 8]
```diff
@@ -216,6 +244,15 @@ impl ConsoleEpollHandler {
                             }
                         }
                     }
+                    CONFIG_EVENT => {
+                        if let Err(e) = self.config_evt.read() {
+                            error!("Failed to get config event: {:?}", e);
+                            break 'epoll;
+                        } else {
+                            (self.interrupt_cb)(&VirtioInterruptType::Config, None).unwrap();
```
Comment (sboeuf) [inline vm-virtio/src/console.rs]: You need to `map_err()` the output.

File: vm-virtio/src/console.rs [hunk 9]
```diff
@@ -365,6 +466,7 @@ impl VirtioDevice for Console {
             };
         self.kill_evt = Some(self_kill_evt);
 
+        let cb = interrupt_cb.clone();
```
Comment (sboeuf) [inline vm-virtio/src/console.rs]: No need to clone the interrupt callback since it should be called once from here. There is no need for the callback to be called from inside the new thread.

File: vm-virtio/src/console.rs [hunk 10]
```diff
@@ -375,12 +477,16 @@ impl VirtioDevice for Console {
...
                 input_evt: self.input.input_evt.try_clone().unwrap(),
+                config_evt: self.input.config_evt.try_clone().unwrap(),
                 kill_evt,
             };
 
             let worker_result = thread::Builder::new()
                 .name("virtio_console".to_string())
-                .spawn(move || handler.run());
+                .spawn(move || {
+                    (cb)(&VirtioInterruptType::Config, None).unwrap();
```
Comment (sboeuf) [inline vm-virtio/src/console.rs]: This should be invoked before to spawn the new thread, avoiding the need for cloning.

File: vmm/Cargo.toml [hunk 11]
```diff
@@ -27,3 +27,11 @@ features = ["elf", "bzimage"]
...
 git = "https://github.com/rust-vmm/vm-memory"
 features = ["backend-mmap"]
+
+[dependencies.lazy_static]
+version = "1.0"
+optional = true
+
+[features]
+default = ["stable"]
+stable = ["lazy_static"]
```
Comment (sboeuf) [inline vmm/Cargo.toml]: I think you should remove this whole declaration and instead add to the existing list of dependencies:
```
[dependencies]
...
kvm-ioctls = { git = "https://github.com/rust-vmm/kvm-ioctls", branch = "master" }
lazy_static = "1.0"
libc = ">=0.2.39"
...
```
The reason is, I don't see why we need to make this dependency optional.

File: vmm/src/lib.rs [hunk 12]
```diff
@@ -2,6 +2,8 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 //
+#[macro_use]
```
Comment (sboeuf) [inline vmm/src/lib.rs]: Leave one blank line between the License header and the first line of code.

File: vmm/src/vm.rs [hunk 13]
```diff
@@ -73,6 +74,11 @@ const KERNEL_64BIT_ENTRY_OFFSET: u64 = 0x200;
 const IOAPIC_RANGE_ADDR: u64 = 0xfec0_0000;
 const IOAPIC_RANGE_SIZE: u64 = 0x20;
 
+// Maximum number of signals
+const MAX_SIGNAL_VAL: usize = 64;
+// Identifier for terminal window size
+const TIOCGWINSZ: u64 = 21523;
```
Comment (sboeuf) [inline vmm/src/vm.rs]: No need to create a constant for this, you should instead use the one from libc. `libc::TIOCGWINSZ`

File: vmm/src/vm.rs [hunk 14]
```diff
@@ -31,7 +31,7 @@ use kvm_bindings::{
 };
 use kvm_ioctls::*;
 use libc::O_TMPFILE;
-use libc::{c_void, siginfo_t, EFD_NONBLOCK};
+use libc::{c_int, c_void, siginfo_t, EFD_NONBLOCK, SA_SIGINFO, SIGHUP, SIGINT, SIGWINCH};
```
Comment (sboeuf) [inline vmm/src/vm.rs]: Based on comment below, please add `TIOCGWINSZ` to the list.

File: vmm/src/vm.rs (hunk 3 above)
Comment (sboeuf) [reply vmm/src/vm.rs]: Well I even think we don't need this list of `present_signals`. As long as you check the signal number the event has been triggered on, and as long as `self.devices.console_input` is not `None`, then you can trigger the console size update.

File: vmm/src/vm.rs [hunk 15]
```diff
@@ -1206,6 +1218,73 @@ impl AsRawFd for EpollContext {
...
+        loop {
+            let mask = MASK.load(Ordering::Relaxed);
+            if mask == 0 {
+                let _ = CVAR.wait(MUTEX.lock().unwrap());
+                thread::yield_now();
+                continue;
+            }
+            signals.clear();
+            if mask & 1 != 0 {
+                signals.push(libc::SIGHUP)
```
Comment (sboeuf) [inline vmm/src/vm.rs]: Please remove all `SIGHUP` and `SIGINT` related code since we're not handling it eventually. And also because if we wanted to handle those signals at some point, it should be done in a separate commit/PR.

File: vmm/src/vm.rs [hunk 16]
```diff
@@ -1616,6 +1720,41 @@ impl<'a> Vm<'a> {
...
+
+                            let ws: WinSize = WinSize {
+                                ws_row: 0,
+                                ws_col: 0,
+                            };
+
+                            unsafe {
+                                libc::ioctl(0, TIOCGWINSZ, &ws);
+                            }
+                            println!("Width:{} Height:{}", ws.ws_col, ws.ws_row);
```
[the remark this answers is not in the document]
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Done!

File: vmm/src/vm.rs [hunk 17]
```diff
@@ -1565,6 +1644,31 @@ impl<'a> Vm<'a> {
         Ok(())
     }
 
+    // Our goal is to separate the handler routine and call it from Vm::start()
```
[the remark this answers is not in the document]
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Done

File: vmm/src/vm.rs (hunk 1 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: As per suggestion from Sebastien, we don't need generic code for signaling. So, these are not relevant now and thus I removed it and associated implementation.

File: vmm/src/vm.rs (hunk 2 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: As per suggestion from Sebastien, we should only do it from vm.rs. So, instead of moving this code to console.rs, I moved everything to vm.rs.

File: vmm/src/vm.rs (hunk 3 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Done according to Sebastien's suggestion.

File: vmm/src/vm.rs (hunk 4 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Explained the action in previous reply.

File: vmm/src/vm.rs (hunk 6 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Explained the action in previous reply.

File: vmm/src/vm.rs (hunk 7 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Explained the action in previous reply. 
(I like the idea of replacing the array with a vector.)

File: vm-virtio/src/console.rs [hunk 18]
```diff
@@ -252,25 +282,75 @@ impl ConsoleInput {
...
+            ws_col: u16, // columns, in characters
+        }
+
+        let ws: WinSize = WinSize {
+            ws_row: 0,
+            ws_col: 0,
+        };
+
+        unsafe {
+            libc::ioctl(0, TIOCGWINSZ, &ws);
```
[the remark this answers is not in the document]
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: We are using signal hook crate as per suggestion from @rbradford  and @sboeuf .

File: vm-virtio/src/console.rs [hunk 19]
```diff
@@ -31,8 +31,27 @@ const INPUT_QUEUE_EVENT: DeviceEventT = 0;
...
+// Console configuration change event is triggered.
+const CONFIG_EVENT: DeviceEventT = 3;
 // The device has been dropped.
-const KILL_EVENT: DeviceEventT = 3;
+const KILL_EVENT: DeviceEventT = 4;
+
+//Console size feature bit
+const VIRTIO_CONSOLE_F_SIZE: u64 = 0;
+// Identifier for terminal window size
+const TIOCGWINSZ: u64 = 21523;
```
[the remark this answers is not in the document]
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: No relevant as we are using signal hook crate.

File: vm-virtio/src/console.rs (hunk 8 above)
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: Done.

File: vm-virtio/src/console.rs (hunk 9 above)
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: Agreed.

File: vm-virtio/src/console.rs (hunk 10 above)
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: Done.

File: vmm/src/vm.rs (hunk 13 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Agreed.

File: vmm/src/vm.rs (hunk 14 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Done.

File: vmm/src/vm.rs (hunk 15 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Done.

File: vmm/src/vm.rs (hunk 5 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: We don't have it anymore.

File: vm-virtio/src/console.rs [hunk 20]
```diff
@@ -31,8 +31,26 @@ const INPUT_QUEUE_EVENT: DeviceEventT = 0;
 const OUTPUT_QUEUE_EVENT: DeviceEventT = 1;
 // Some input from the VMM is ready to be injected into the VM.
 const INPUT_EVENT: DeviceEventT = 2;
+// Console configuration change event is triggered.
+const CONFIG_EVENT: DeviceEventT = 3;
 // The device has been dropped.
-const KILL_EVENT: DeviceEventT = 3;
+const KILL_EVENT: DeviceEventT = 4;
```
Comment (rbradford) [inline vm-virtio/src/console.rs]: Would have been better to make CONFIG_EVENT  =4

File: vmm/src/vm.rs [hunk 21]
```diff
@@ -490,6 +492,29 @@ impl UserIoapicIrq {
         UserIoapicIrq { ioapic, irq }
     }
 }
+struct WinSize {
```
Comment (rbradford) [inline vmm/src/vm.rs]: You can get rid of this top-level struct.

File: vmm/src/vm.rs [hunk 22]
```diff
@@ -490,6 +492,29 @@ impl UserIoapicIrq {
...
+impl WinSize {
+    fn new() -> WinSize {
+        WinSize {
+            ws_row: 0,
+            ws_col: 0,
+        }
+    }
+
+    fn get_win_size() -> (u16, u16) {
+        let ws: WinSize = WinSize::new();
```
Comment (rbradford) [inline vmm/src/vm.rs]: Just do let `ws: struct {rows: u16, cols: u16} = {rows: 0, cols: 0};`

File: vm-virtio/src/console.rs [hunk 23]
```diff
@@ -252,25 +283,61 @@ impl ConsoleInput {
         in_buffer.extend(input);
         let _ = self.input_evt.write(1);
     }
+
+    pub fn update_console_size(&self, cols: u16, rows: u16) {
+        self.config.lock().unwrap().update_console_size(cols, rows);
```
Comment (rbradford) [inline vm-virtio/src/console.rs]: Check that `VIRTIO_F_CONSOLE_F_SIZE` is in the acked_features before doing anything in this function.

File: vm-virtio/src/console.rs [hunk 24]
```diff
@@ -331,12 +398,33 @@ impl VirtioDevice for Console {
...
+        if let Some(end) = offset.checked_add(data.len() as u64) {
+            // This write can't fail, offset and end are checked against config_len.
+            data.write_all(&config_slice[offset as usize..cmp::min(end, config_len) as usize])
+                .unwrap();
+        }
     }
 
-    fn write_config(&mut self, _offset: u64, _data: &[u8]) {
-        warn!("Device specific configuration is not defined yet");
+    fn write_config(&mut self, offset: u64, data: &[u8]) {
```
Comment (rbradford) [inline vm-virtio/src/console.rs]: Looking at the spec I think the fields are completely read-only from the driver site.

File: vmm/src/vm.rs [hunk 25]
```diff
@@ -1576,6 +1614,16 @@ impl<'a> Vm<'a> {
...
         vcpu_thread_barrier.wait();
 
+        if let Some(console_input) = &self.devices.console_input {
+            let console_input_clone = console_input.clone();
+            let signals = Signals::new(&[SIGWINCH]);
+            match signals {
+                Ok(sig) => {
+                    thread::spawn(move || Vm::handler(sig, console_input_clone));
+                }
+                Err(_) => error!("Signal not found"),
```
Comment (rbradford) [inline vmm/src/vm.rs]: Maybe use the actual error returned.

File: vmm/src/vm.rs [hunk 26]
```diff
@@ -1525,6 +1554,15 @@ impl<'a> Vm<'a> {
         Ok(())
     }
 
+    fn handler(signals: Signals, console_input_clone: Arc<vm_virtio::ConsoleInput>) {
```
Comment (rbradford) [inline vmm/src/vm.rs]: Please give this function a slightly more descriptive name.

File: vm-virtio/src/console.rs [hunk 27]
```diff
@@ -31,8 +31,26 @@ const INPUT_QUEUE_EVENT: DeviceEventT = 0;
...
 const INPUT_EVENT: DeviceEventT = 2;
+// Console configuration change event is triggered.
+const CONFIG_EVENT: DeviceEventT = 3;
 // The device has been dropped.
-const KILL_EVENT: DeviceEventT = 3;
+const KILL_EVENT: DeviceEventT = 4;
+
+//Console size feature bit
+const VIRTIO_CONSOLE_F_SIZE: u64 = 0;
+// Identifier for terminal window size
```
Comment (rbradford) [inline vm-virtio/src/console.rs]: Dunno what this comment is for.

File: vm-virtio/src/console.rs (hunk 20 above)
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: Done.

File: vmm/src/vm.rs (hunk 22 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: I am not sure if this is a valid syntax. But, the following works.
struct WS {rows: u16, cols: u16};
let ws: WS = WS {rows: 0u16, cols: 0u16};

File: vm-virtio/src/console.rs (hunk 23 above)
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: Done.

File: vmm/src/vm.rs (hunk 22 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: I am not sure if this is a correct syntax. The following works though.

struct WS {rows: u16, cols: u16};
let ws: WS = WS {rows: 0u16, cols: 0u16};

File: vmm/src/vm.rs (hunk 25 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Printed the error along with this message.

File: vmm/src/vm.rs (hunk 26 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Done.

File: vmm/src/vm.rs (hunk 21 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: Done.

File: vm-virtio/src/console.rs (hunk 24 above)
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: From the the console size perspective, Yes. However, for console size we have exposed virtio_console_config struct, where emerg_wr is a part of it. In future, we plan to enable emerg_wr for which we may need write_config. Therefore, I kept it implemented.

File: vm-virtio/src/console.rs (hunk 27 above)
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: Cleaned.

File: src/main.rs [hunk 28]
```diff
@@ -616,6 +635,21 @@ mod tests {
                 .is_empty())
         }
 
+        fn get_win_size(&self) -> (u16, u16) {
+            let col = self
+                .ssh_command("echo $COLUMNS >  winsize; cat winsize")
```
Comment (rbradford) [inline src/main.rs]: Why do you have to redirect it to a file only to cat it?

File: vm-virtio/src/console.rs (hunk 24 above)
Comment (rbradford) [reply vm-virtio/src/console.rs]: You should delete it for now as we aren't advertising that feature (and i'm not convinced it's a very useful one for us). Carrying some code that is unused isn't helpful to us.

File: vmm/src/vm.rs [hunk 29]
```diff
@@ -491,6 +493,22 @@ impl UserIoapicIrq {
     }
 }
 
+pub fn get_win_size() -> (u16, u16) {
+    struct WS {
```
Comment (rbradford) [inline vmm/src/vm.rs]: You need to `#repr(C)` this.

Comment (rbradford) [reply vmm/src/vm.rs]: Might be simpler to use a "tuple struct"

File: vm-virtio/src/console.rs (hunk 24 above)
Comment (fazlamehrab) [reply vm-virtio/src/console.rs]: Ok.

File: vmm/src/vm.rs (hunk 29 above)
Comment (fazlamehrab) [reply vmm/src/vm.rs]: How to return a 'tuple' from 'tuple struct'?
I used  `#[repr(C)]` before the struct definition.

File: src/main.rs (hunk 28 above)
Comment (fazlamehrab) [reply src/main.rs]: I just realized that I can just use pipe. Thanks.

File: vmm/src/vm.rs (hunk 29 above)
Comment (rbradford) [reply vmm/src/vm.rs]: > I have used #[repr(C)] before the struct definition.

I don't see this.

Comment (fazlamehrab) [reply vmm/src/vm.rs]: Could you please check now?

File: vmm/src/lib.rs (hunk 12 above)
Comment (fazlamehrab) [reply vmm/src/lib.rs]: Done!

File: src/main.rs [hunk 30]
```diff
@@ -186,11 +189,39 @@ mod tests {
...
+        struct WS {
+            rows: u16,
+            cols: u16,
+        };
+        let ws: WS = WS {
+            rows: 0u16,
+            cols: 0u16,
+        };
+        unsafe {
+            ioctl(0, TIOCGWINSZ, &ws);
```
Comment (rbradford) [inline src/main.rs]: You might want to result a result of a vmm_sys_util::errno::Error generated by `Error::last()`

I think the ioctl might be failing on the Jenkins as the tests are not run inside a tty.

File: src/main.rs [hunk 31]
```diff
@@ -260,6 +260,7 @@ extern crate lazy_static;
 #[cfg(test)]
 #[cfg(feature = "integration_tests")]
 mod tests {
+
```
Comment (rbradford) [inline src/main.rs]: You should drop the whitespace changes in this file.

File: Cargo.toml [hunk 32]
```diff
@@ -9,6 +9,7 @@ clap = "2.33.0"
 lazy_static = "1.3.0"
 log = { version = "0.4.8", features = ["std"] }
 vmm = { path = "vmm" }
+libc = "0.2.60"
```
Comment (rbradford) [inline Cargo.toml]: Is this still needed now you removed the tests?

File: src/main.rs (hunk 31 above)
Comment (fazlamehrab) [reply src/main.rs]: Sure.

File: Cargo.toml (hunk 32 above)
Comment (fazlamehrab) [reply Cargo.toml]: Not needed anymore.

**Comment by fazlamehrab** [comment]
> I thought you were going to keep the test for the sizing but mark it as `#[ignore]` so that it could still be run manually?

@rbradford There is no difference between running tests in Jenkins and doing it manually in the local machine in terms of interactive window because none of them have it. Therefore, there is no use of those tests.
